lib/LINZ/BdeUpload.pm

```diff
@@ -369,6 +369,58 @@ use fields qw{ cfg db repository tables tmp_base tmp fid current_dataset
 
 our $tmp_prefix='tmp_bde_upload_';
 
+sub uploadDataToTempTable
+{
+    my($self,$tablename,$tmpfile,$columns) = @_;
+
+    $self->db->_setDbMessageHandler;
+
+    DEBUG('Loading file ' . $tmpfile . ' into table ' . $tablename);
+
+    my $uploadId = $self->db->uploadId;
+    my $dbh = $self->db->_dbh;
+
+    my @quotedColumns;
+    foreach my $col (split(/\|/, $columns)) {
+        push(@quotedColumns, $dbh->quote_identifier( $col ));
+    }
+
+    #DEBUG('quoted columns are ' . join(',',@quotedColumns));
+
+    my $sql = 'SELECT bde_control._bde_RefreshLock(?)';
+    $dbh->do( $sql, {}, $uploadId);
+
+    $sql = 'SELECT bde_control._bde_WorkingCopyTableOid(?,?)';
+    my ($tmptable) = $dbh->selectrow_array( $sql, {}, $uploadId, $tablename);
+
+    DEBUG('Temp table is ' . $tmptable);
+    if ( ! $tmptable ) {
+        ERROR 'Cannot load file ' . $tmpfile
+          . ' into table ' . $tablename
+          . ' as working copy of table does not exist';
+        $self->db->_clearDbMessageHandler;
+        return 0;
+    }
+
+    $sql = 'LOCK TABLE ' . $tmptable . ' IN ACCESS EXCLUSIVE MODE';
+    $dbh->do( $sql );
+
+    $sql = 'COPY ' . $tmptable . ' (' . join(',', @quotedColumns)
+         . ") FROM stdin WITH DELIMITER '|' NULL AS ''";
```

Comment:
What's the performance difference on a local machine between STDIN and a direct file input with the SQL COPY command? If it's a large enough amount, then maybe calling this new version of uploadDataToTempTable should be configurable?

Reply:
With current code this is the comparison:

```
# local datbase, full dataset (20170506184000 level 0 update)
# time sudo -Eu bde /opt/bde-processor/bde_upload btl /etc/linz-bde-uploader/linz_bde_uploader --rebuild crs_coordinate 2>&1 | tee /tmp/log


# COPY FROM /path/to/file:

(1st)
real    13m55.150s -- enables versions...
user    0m58.702s
sys     0m10.871s

(2nd)
real    4m13.258s
user    0m59.761s
sys     0m9.038s

(3rd)
real    4m10.152s
user    0m59.428s
sys     0m9.601s

# COPY FROM stdin:

(4th)
real    4m26.515s
user    1m3.206s
sys     0m19.205s

(5th)
real    4m27.438s
user    1m3.398s
sys     0m18.789s


```

So it looks to me that COPY FROM STDIN consumes double `sys` calls, which are probably unoptimized reads (we read each line and send it over the wire, rather than doing it in block) but still the overall time is not that different (6% slower). I think it can be optimized by reading/sending things in blocks. In any case I don't think it's useful for it to be configured as even w/out optimizing this I'm sure it'll become faster once `bde_copy` is made to output directly to stdout as we'd save a lot of writes to disk for the creation of the temporary file.

Reply:
Great!

Reply:
Reading in chunks doesn't change the situation much:
```
XXX put 387225 chunks of buffers to copydata -- sys     0m18.428s
XXX put 38723 chunks of buffers to copydata -- sys     0m17.886s
```

So I guess pg_putcopydata already creates a buffer internally and perl already buffers data upon calling the line-oriented reads, and the double sys calls are because of the added network writes, not present when the server reads the file locally.

Making use of streaming vs. from-file configurable would mean having to handle the configuration even when `bde_copy` is capable of outputting to stdout, in whcih case the configuration would be useless, I would not go there.

So enough said about performance, I'll move to the private APIs issue.

Reply:
So, beside pg-schema private function this code is also having BdeUpload.pm call BdeDatabase private functions _setDbMessageHandler and _clearDbMessageHandler - would it make sense to move the whole `uploadDataToTempTable` implementation from BdeUpload.pm to BdeDatabase ? Would also make BdeUpload.pm untouched